### PR TITLE
Batch Gift Entry: Elevate widget updates

### PR DESCRIPTION
**Pull request by psadie-sfdo**
WI: [W-8970003](https://gus.lightning.force.com/lightning/_classic/%2Fa07B0000008x92mIAA)
WI: [W-8809426](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000009535oIAA/view)

Batch Gift Entry now supports loading of the payment widget and handling of authorized, captured and declined credit card gifts

Custom Labels:
- commonEditPaymentInformation
- commonCardLast4

# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: src/lwc/geFormWidgetTokenizeCard/__tests__/geFormWidgetTokenizeCard.test.js [hunk 1]
```diff
@@ -2,6 +2,15 @@ import { createElement } from 'lwc';
...
 import { DISABLE_TOKENIZE_WIDGET_EVENT_NAME } from 'c/geConstants';
 import { fireEvent } from 'c/pubsubNoPageRef';
+import { getRecord } from 'lightning/uiRecordApi';
+import { getObjectInfo } from 'lightning/uiObjectInfoApi';
+import { registerLdsTestWireAdapter } from '@salesforce/sfdx-lwc-jest';
+
+const mockGetRecord = require('./data/DIMockRecord.json');
+const mockObjectInfo = require('./data/dataImportObjectDescribeInfo.json');
+
+const getRecordAdapter = registerLdsTestWireAdapter(getRecord);
```
Comment (bdvorachek) [inline src/lwc/geFormWidgetTokenizeCard/__tests__/geFormWidgetTokenizeCard.test.js:12]: Both getRecordAdapter and getObjectInfoAdapter are no longer needed as of sfdx-lwc-jest 0.12, can delete these two lines.

File: src/lwc/geFormWidgetTokenizeCard/__tests__/geFormWidgetTokenizeCard.test.js [hunk 2]
```diff
@@ -128,6 +145,94 @@ describe('c-ge-form-widget-tokenize-card', () => {
...
+            CAPTURED: 'CAPTURED'
+        }
+        element.widgetDataFromState = {
+            ...element.widgetDataFromState,
+            [PAYMENT_METHOD_FIELD]: CREDIT_CARD,
+            [DATA_IMPORT_PARENT_BATCH_LOOKUP]: 'DUMMY_ID',
+            [DATA_IMPORT_PAYMENT_STATUS]: 'CAPTURED'
+        };
+
+        getObjectInfoAdapter.emit(mockObjectInfo);
```
Comment (bdvorachek) [inline src/lwc/geFormWidgetTokenizeCard/__tests__/geFormWidgetTokenizeCard.test.js:179]:
```suggestion
        getObjectInfo.emit(mockObjectInfo);
```
Can emit directly to the imported wire adapter. Check for other references to the (now deleted) testAdapters, and use the adapter directly.

File: src/lwc/geFormWidgetTokenizeCard/__tests__/geFormWidgetTokenizeCard.test.js [hunk 3]
```diff
@@ -128,6 +145,94 @@ describe('c-ge-form-widget-tokenize-card', () => {
...
+        getRecordAdapter.emit(mockGetRecord);
```
Comment (bdvorachek) [inline src/lwc/geFormWidgetTokenizeCard/__tests__/geFormWidgetTokenizeCard.test.js:180]:
```suggestion
        getRecord.emit(mockGetRecord);
```

**Review by bdvorachek — CHANGES_REQUESTED** [review]

File: src/lwc/geFormWidgetTokenizeCard/__tests__/geFormWidgetTokenizeCard.test.js [hunk 4]
```diff
@@ -2,6 +2,15 @@ import { createElement } from 'lwc';
 import GeFormWidgetTokenizeCard from 'c/geFormWidgetTokenizeCard';
 import { DISABLE_TOKENIZE_WIDGET_EVENT_NAME } from 'c/geConstants';
 import { fireEvent } from 'c/pubsubNoPageRef';
+import { getRecord } from 'lightning/uiRecordApi';
+import { getObjectInfo } from 'lightning/uiObjectInfoApi';
+import { registerLdsTestWireAdapter } from '@salesforce/sfdx-lwc-jest';
```
Comment (bdvorachek) [inline src/lwc/geFormWidgetTokenizeCard/__tests__/geFormWidgetTokenizeCard.test.js:7]: This import can get removed along with the other suggested changes.

File: src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js [hunk 5]
```diff
@@ -1,112 +1,287 @@
...
+    }
+
+    hasReadOnlyStatus() {
+        if(!this.isInBatchGiftEntry() || isEmpty(this.paymentTransactionStatusValues)) {
+            return false;
+        }
+        return this.readOnlyStatuses().includes(this.currentPaymentStatus());
     }
 
     hasValidPaymentMethod(paymentMethod) {
```
Comment (bdvorachek) [inline src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js:281]:
```suggestion
    hasValidPaymentMethod() {
```

File: src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js [hunk 6]
```diff
@@ -1,112 +1,287 @@
...
+        if (this.isInBatchGiftEntry() && paymentMethod === PAYMENT_METHODS.ACH) {
```
Comment (bdvorachek) [inline src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js:282]:
```suggestion
        if (this.isInBatchGiftEntry() && this._currentPaymentMethod === PAYMENT_METHODS.ACH) {
```

File: src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js [hunk 7]
```diff
@@ -1,112 +1,287 @@
...
+            return false;
+        }
         return paymentMethod === PAYMENT_METHODS.ACH
             || paymentMethod === PAYMENT_METHOD_CREDIT_CARD;
```
Comment (bdvorachek) [inline src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js:286]:
```suggestion
        return this._currentPaymentMethod === PAYMENT_METHODS.ACH
            || this._currentPaymentMethod === PAYMENT_METHOD_CREDIT_CARD;
```

File: src/lwc/geFormRenderer/geFormRenderer.html [hunk 8]
```diff
@@ -101,20 +101,20 @@
                             </c-util-illustration>
                         </div>
                     </template>
-                    <template if:true={isSingleGiftEntry}>
-                        <template if:true={showElevateTransactionWarning}>
-                            <div class="slds-notify slds-notify_alert payment-warn" role="alert">
-                                <span class="slds-assistive-text">{CUSTOM_LABELS.commonWarning}</span>
-                                <span class="slds-icon_container slds-icon-utility-warning slds-m-right_x-small">
+                    <template if:true={showElevateTransactionWarning}>
```
[the remark this answers is not in the document]
Comment (voduyemi) [reply src/lwc/geFormRenderer/geFormRenderer.html:104]: Yup!

**Review by bdvorachek — APPROVED** [review]

**Review by dfuller114 — APPROVED** [review]
Added a general comment that is non-blocking for how we manage state in geFormWidgetTokenizeCard which encompasses several other comments I would make in that .js file. I think we should look into refactoring how we manage the different widget states soon so that understanding how the different states affect the rest of the form and the widget itself doesn't become harder to understand and maintain. Otherwise, everything else looks good to me!

File: src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js [hunk 9]
```diff
@@ -1,114 +1,289 @@
...
+    dataImportId;
     _currentPaymentMethod = undefined;
-    _hasPaymentMethodInTemplate = false;
+    _isReadOnlyMode = false;
+    _showCancelButton = false;
+    _cardLast4;
+    _cardExpirationDate;
+    _widgetDataFromState;
+
+    setMode (mode) {
```
[the remark this answers is not in the document]
Comment (voduyemi) [reply src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js:66]: Absolutely!

File: src/lwc/geGiftEntryFormApp/geGiftEntryFormApp.html [hunk 10]
```diff
@@ -7,7 +7,6 @@
                 onedit={handleEditBatch}
                 onprocessbatch={handleProcessBatch}>
         </c-ge-batch-gift-entry-header>
-        <br/>
```
Comment (ptewson-sfdo) [inline src/lwc/geGiftEntryFormApp/geGiftEntryFormApp.html:10]: :tada:

File: src/labels/CustomLabels.labels [hunk 11]
```diff
@@ -5519,6 +5519,14 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
         <shortDescription>Edit</shortDescription>
         <value>Edit</value>
     </labels>
+    <labels>
+        <fullName>commonEditPaymentInformation</fullName>
```
Comment (voduyemi) [inline src/labels/CustomLabels.labels:5523]: @melissabarber Updated label here

File: src/labels/CustomLabels.labels [hunk 12]
```diff
@@ -5519,6 +5519,14 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
...
+        <categories>Recurring-Donations, Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Text to edit elevate payment information</shortDescription>
+        <value>Edit Payment Information</value>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5528]:
```suggestion
        <value>Edit payment information.</value>
```

File: src/labels/CustomLabels.labels (hunk 11 above)
Comment (melissabarber) [reply src/labels/CustomLabels.labels:5523]: Made one suggestion. After that's committed, this looks good to go.

**Review by sokosaurus — APPROVED** [review]

**Pull request merged**